Feature/open library

juanmbellini opened the pull request:
# Summary

This PR adds an integration with Open Library, according to the trello card: [https://trello.com/c/APRqeorN](https://trello.com/c/APRqeorN).

# Notes

 Instead of creating a DTO to communicate data between our platform and Open Library, this implementation uses the Jackson's JsonNode API, as the returned JSON returned by Open Library is very complex, and mapping it into a POJO would create a very complex object.

# Trello Card

https://trello.com/c/APRqeorN

juanmbellini committed:
Add a search by isbn resource

- Include a search by isbn method in the BookRepository
- Also specify a regex for all the API routes (this is necessary to disambiguate the "find by id" and "find by isbn" routes)

juanmbellini committed:
Create the OpenLibraryService

- Create the service as an interface
- Inclde an implementation using RestTemplate
- Make use of the OpenLibraryService from the BookController, storing the Book when it is retrieved

juanmbellini committed:
Add configuraiton properties to the application.yml file

- Include metadata and spring-boot-configuration-processor

juanmbellini committed:
Implement the integration with Open Library

- Mapping of Books is performed using Jackson's JsonNode API (this avoids creating a very complex POJO that matches the Open Library book model)

juanmbellini committed:
Add a helper method to the BookRepository to search by authors

- The Open Library integration saves Books with more than one author as a string, appending author names with ", ". This commit adds the ability to search by a List of authors using the same logic as the integration implementation

juanmbellini committed:
Add an OpenLibraryService mock to the BookControllerTest

darkBasic666 commented on src/main/java/wolox/training/services/open_library/RestTemplateOpenLibraryService.java:
```diff
@@ -0,0 +1,100 @@
+                        .map(n -> n.path("name"))
+                        .map(JsonNode::asText)
+                );
+                final var publisher = Utils.sortAndJoinAuthors(
```
Utils.sortAndJoinAuthors  not represent what  do . maybe using  more abstract name to this funtion, aclarify what to do.

juanmbellini replied:
Okey, I can call it `sortAndJointWithComma` or something like that.

darkBasic666 commented on src/test/java/wolox/training/utils/ValuesGenerator.java:
```diff
@@ -24,7 +24,7 @@ private ValuesGenerator() {
      */
     public static long validBookId() {
-        return Faker.instance().number().numberBetween(Long.MIN_VALUE, Long.MAX_VALUE);
+        return Faker.instance().number().numberBetween(1, Long.MAX_VALUE);
```
What represents this '1'? 
Maybe moving to constant clarify this number represents.

juanmbellini replied:
Okey, I'll extract it into a constant. Its just a minimum number for ids. It was changed to 1 because postgres'  `bigserial` type has values between 1 and `Long.MAX_VALUE` (https://www.postgresql.org/docs/10/datatype-numeric.html). Also, as the `BookController` adds the regex for the mapping of "get by id" and "delete by id" methods, using the `\\d+` regex, if the returned value of this method is negative, the tests start to fail.

juanmbellini committed:
Add fixes according to comments

- Change sortAndJoinAuthors to sortAndJoinWithCommas
- Add a constant indicating what does the 1 means in validBookId and validUserId
